src/commcare_cloud/commcare_cloud.py

```diff
@@ -150,9 +150,70 @@ def make_parser(available_envs, formatter_class=RawTextHelpFormatter,
     return parser, subparsers, commands
 
 
+def compile_changelog():
+    # Parse the contents of the changelog dir
+    changelog_contents = []
+    files_to_ignore = ['0000-changelog.md', 'index.md']
+    changelog_dir = os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..', 'docs', 'changelog')
+    sorted_files = _sort_files(changelog_dir)
+    for change_file_name in sorted_files:
+        if change_file_name not in files_to_ignore:
+            _parse_changelog_file(changelog_contents, changelog_dir, change_file_name)
+    return changelog_contents
+
+
+def _parse_changelog_file(changelog_contents, changelog_dir, change_file_name):
+    try:
+        with open(os.path.join(changelog_dir, change_file_name)) as change_file:
+            change_context = ''
+            change_date = ''
+            in_change_context = False
+            change_action_required = False
+            reached_details_line = False
+            for line_number, line in enumerate(change_file):
+                if line_number == 0:
+                    change_summary = re.search('(?<=\.).*', line).group().strip()
+                if '**Date:**' in line:
+                    change_date = line.split('**Date:**')[1].strip()
+                if '**Optional per env:**' in line:
+                    option = line.split('**Optional per env:**')[1].strip().lower()
+                    if "no" in option:
+                        change_action_required = True
+                if '## Details' in line:
+                    in_change_context = False
+                    reached_details_line = True
+                if in_change_context:
+                    change_context += line.replace('\n', '')
+                if '## Change Context' in line:
+                    in_change_context = True
+            assert change_file_name and change_context and change_date and change_summary and \
+                   change_action_required and reached_details_line
+            this_changelog = {'filename': change_file_name,
+                              'context': change_context,
+                              'date': change_date,
+                              'summary': change_summary,
+                              'action_required': change_action_required}
+
+        changelog_contents.append(this_changelog)
+    except (AttributeError, AssertionError):
+        print("Error parsing the file {}.".format(change_file_name))
+        sys.exit(1)
```

Comment:
There are a couple try/except antipatterns going on here as I see it:
- `try`/`except` blocks with a large (many-line) `try` block—ideally you just have a few lines in the `try` block so that it's quite clear what errors you're planning to catch
- Catching a very general error type (`AttributeError`), which could result in catching errors other than the ones you had in mind
- Catching an error and exiting when simply _not_ handling the error would do just as well or better
- Blanket suppression of error messages

That last one is practically the most important I think. The point of the `try`/`except` here as I understand it is basically to catch any type of error that might happen so it doesn't bubble up (i.e. `except Exception` would be the more explicit way to do this). But why don't we want the error to bubble up? Let's say I just added a new file, and I run `make` (or I forget to run `make` and then see this error in travis), and then all I see is `Error parsing the file docs/changelog/foo.md`. What's my next step? I'll probably search the code for that, and then remove the `try`/`except` here and run it again so that I can see the actual error! Why not just remove the try except to begin with?

One answer I can think of is that then the error output might not include the name of the problematic file. The solution to this would be to do a blanket catch (`except Exception`) print out the message like you do here, but then have a bare `raise` statement (to reraise the previous error) rather than exit. That way they see your message _and_ the traceback that'll help them identify the exact problem quicker.

To avoid the long `try` block, I'd suggest removing it here entirely, but then above where this function is called have

```
try:
    _parse_changelog_file(changelog_contents, changelog_dir, change_file_name)
except Exception:
    print("Error parsing the file {}.".format(change_file_name))
    raise
```

Reply:
https://github.com/dimagi/commcare-cloud/pull/1979/commits/2e3be14fcdaa6f31038d928256d52baa33f630ef

One thing I noticed with this:
The printout only shows up when running `manage-commcare-cloud make-changelog-index`, and not when running `make` (because in that case, the printout is part of the output, so it gets written to `index.md` ☹️ Is that acceptable? I would imagine the workflow is to run the `manage-commare-cloud` command before actually running make, right?)